lms/djangoapps/course_structure_api/v0/views.py

```diff
@@ -290,3 +297,362 @@ class CourseGradingPolicy(CourseViewMixin, ListAPIView):
+
+            # if a parent block was provided, update this block's data based on the parent's data
+            if parent_block_info:
+                self.depth = parent_block_info.depth + 1
+                self.descendants_of_parent = parent_block_info.descendants_of_self
+
+    @view_course_access(depth=None)
+    def list(self, request, course, return_blocks=True, return_nav=True, *args, **kwargs):
+        """
+        REST API endpoint for listing all the blocks and/or navigation information in the course,
+        while regarding user access and roles.
+        """
+        # set starting point
+        start_block = course
+
+        # initialize request and result objects
+        request_info = self.RequestInfo(request, course)
+        result_info = self.ResultInfo(return_blocks, return_nav)
+
+        # create and populate a field data cache by pre-fetching for the course (with depth=None)
+        request_info.field_data_cache = FieldDataCache.cache_for_descriptor_descendents(
+            course.id, request.user, course, depth=None,
+        )
+
+        # start the recursion with the start_block
+        self.recurse_blocks_nav(request_info, result_info, self.BlockInfo(start_block))
+
+        # return response
+        response = {"root": unicode(start_block.location)}
+        result_info.update_response(response, return_blocks, return_nav)
+        return Response(response)
+
+    def recurse_blocks_nav(self, request_info, result_info, block_info):
+        """
+        A depth-first recursive function that supports calculation of both the list of blocks in the course
+        and the navigation information up to requested navigation_depth of the course.
+
+        Arguments:
+        """
+        # bind user data to the block
+        block_info.block = get_module_for_descriptor(
+            request_info.request.user,
+            request_info.request,
+            block_info.block,
+            request_info.field_data_cache,
+            request_info.course.id
+        )
+
+        # verify the user has access to this block
+        if not has_access(request_info.request.user, 'load', block_info.block, course_key=request_info.course.id):
+            return
+
+        # set basic information about the block
+        self.set_block_value(request_info, result_info, block_info)
+
+        # descendants
+        self.update_descendants(request_info, result_info, block_info)
+
+        # children: recursively call the function for each of the children, while supporting dynamic children.
+        children = []
+        if block_info.block.has_children:
+            children = get_dynamic_descriptor_children(block_info.block, request_info.request.user.id)
+            for child in children:
+                self.recurse_blocks_nav(request_info, result_info, self.BlockInfo(child, block_info))
+            if request_info.children:
+                block_info.value["children"] = [unicode(child.location) for child in children]
```

Comment:
Suggestion: save a pass through children by building this list within the for loop? This would only really matter if len(children) is generally very large